### PR TITLE
Namespace save data

### PR DESCRIPTION
**Pull request by AydinHassan**
I had to change the save file, if we kept it the same, each time you would run an old workshop, it would overwrite the new save data :(.

For now, we migrate this old file when we are in that workshop and it has been updated, then we remove it.

### CODE REVIEW

File: src/UserStateSerializer.php [hunk 1]
```diff
@@ -37,60 +57,70 @@ public function __construct($path)
...
     {
-        return file_put_contents($this->path, json_encode([
-            'completed_exercises'   => $state->getCompletedExercises(),
-            'current_exercise'      => $state->getCurrentExercise(),
-        ]));
+        $saveFile = sprintf('%s/%s', $this->path, static::SAVE_FILE);
+
+        if (!file_exists($saveFile)) {
+            $data = [];
+            $data[$this->workshopName] = [
```
Comment (mikeymike) [inline src/UserStateSerializer.php]: this line and the one in the else is the same... maybe just do the initialisation of $data in a short ternary ?

Comment (AydinHassan) [inline src/UserStateSerializer.php]: Yeah good shout!

Comment (AydinHassan) [inline src/UserStateSerializer.php]: You fucking love ternaries.

File: src/UserStateSerializer.php [hunk 2]
```diff
@@ -100,10 +125,79 @@ public function deSerialize()
...
+    }
+
+    /**
+     * @param string $filePath
+     * @return array|null
+     */
+    private function readJson($filePath)
+    {
+        if (!file_exists($filePath)) {
+            return null;
```
Comment (mikeymike) [inline src/UserStateSerializer.php]: is there any benefit to returning null rather than an empty array here?

**Comment by mikeymike** [comment]
overall LGTM dude... complex beast but yeah I'm good with it

File: src/UserStateSerializer.php (hunk 2 above)
Comment (AydinHassan) [inline src/UserStateSerializer.php]: Not really sure - but null could be a corrupted file, so it needs to be removed, whereas empty array is valid json really.

**Comment by AydinHassan** [comment]
Yeah I feel you, it felt gross writing it :(

**Pull request merged**